app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -3,8 +3,59 @@ kind: registration
 name: Global Registration
 model: ProvisioningTemplate
 -%>
+#!/bin/sh
+<%
+    curl_opts = ["-H 'Content-Type: application/json'", "-H 'Accept: application/json'", "-H 'Authorization: Bearer #{@auth_token}'"]
+%>
+
+# Template parameters:
+<%= "#  - User: [#{@user.login}]" if @user %>
+<%= "#  - Organization: [#{@organization.name}]" if @organization %>
+<%= "#  - Location: [#{@location.name}]" if @location %>
+<%= "#  - Hostgroup: [#{@hostgroup.name}]" if @hostgroup %>
 
 if ! [ $(id -u) = 0 ]; then
     echo "Please run as root"
     exit 1
 fi
+
+if [ -f /etc/os-release ] ; then
+  . /etc/os-release
+fi
+
+HOSTNAME=$(hostname -f)
+
+echo "#"
+echo "# Running registration"
+echo "# Hostname: $HOSTNAME"
+echo "# OS: $NAME"
+echo "#"
+
+if [ $ID = rhel ] ; then
+<% if plugin_present?('katello') %>
+    CONSUMER_RPM=$(mktemp --suffix .rpm)
+
+    curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
+    yum localinstall $CONSUMER_RPM -y
+    yum install subscription-manager -y
+    subscription-manager register --org=<%= @organization.label %> --activationkey="sample-key"
+
+    rm -f $CONSUMER
+<% else %>
+    curl -X POST <%= foreman_server_url %>/api/hosts \
+     <%= curl_opts.join(' ') %> \
+     -d '{ "host": { "name": "'$HOSTNAME'", "build": "false", "managed": "false"
+     <%= ", \"organization_id\": \"#{@organization.id}\"" if @organization %>
+     <%= ", \"location_id\": \"#{@location.id}\"" if @location %>
+     <%= ", \"hostgroup_id\": \"#{@hostgroup.id}\"" if @hostgroup %>
+     }}'
+<% end %>
+else
+    curl -X POST <%= foreman_server_url %>/api/hosts \
+     <%= curl_opts.join(' ') %> \
+     -d '{ "host": { "name": "'$HOSTNAME'", "build": "false", "managed": "false"
+     <%= ", \"organization_id\": \"#{@organization.id}\"" if @organization %>
+     <%= ", \"location_id\": \"#{@location.id}\"" if @location %>
+     <%= ", \"hostgroup_id\": \"#{@hostgroup.id}\"" if @hostgroup %>
+     }}'
+fi
```

Comment:
This is getting out of hand, how about

```shell
HAS_KATELLO=<%= plugin_present?('katello') ? 1 : 0 %>

if [ $HAS_KATELLO -eq 1 ] && [ $ID = rhel ]; then
  # subman stuff
else
  # curl-based registration
fi

```

Reply:
Well I wanted to do it this way too but here is the problem: Foreman doesn't know Katello's stuff, so with your changes if you try to render the template you will get error:
```
Safemode doesn't allow to access 'subscription_manager_configuration_url'
```
and:
```
Safemode doesn't allow to access organization.label
```

Reply:
So how about you turn the curl based registration either into a snippet at the template level or turn it into a function inside the template and then just include/call it at all the various places?

Reply:
Created new snippet `create_host_curl`.